eth/sync.go

```diff
@@ -171,7 +171,11 @@ func (pm *ProtocolManager) synchronise(peer *peer) {
 	td := pm.blockchain.GetTd(currentBlock.Hash(), currentBlock.NumberU64())
 
 	pHead, pTd := peer.Head()
-	if pTd.Cmp(td) <= 0 {
+	if cmp := pTd.Cmp(td); cmp <= 0 {
+		if cmp < 0 {
+			// propagate our better chain back to peers
+			go pm.BroadcastBlock(currentBlock, true)
```

Comment:
This is a big no-no. It propagates entire blocks to everyone because one peer is out of sync. I don't think we should "think" in place of our peers. We announce in the handshake what our TD is and also do it on every block.

Reply:
If the block is known by a peer, BroadcastBlock will skip them, so I don't think it's a problem. Beside, another solution is to broadcast to only that peer instead of all peers.